include/irr/asset/ICPUVirtualTexture.h

```diff
@@ -0,0 +1,337 @@
+    }
+
+    STextureData pack(const ICPUImage* _img, const IImage::SSubresourceRange& _subres, ISampler::E_TEXTURE_CLAMP _wrapu, ISampler::E_TEXTURE_CLAMP _wrapv, ISampler::E_TEXTURE_BORDER_COLOR _borderColor) override
+    {
+        const E_FORMAT format = _img->getCreationParameters().format;
+        uint32_t smplrIndex = 0u;
+        ICPUVTResidentStorage* storage = nullptr;
+        {
+            auto found = m_storage.find(getFormatClass(format));
+            if (found==m_storage.end())
+                return STextureData::invalid();
+            storage = static_cast<ICPUVTResidentStorage*>(found->second.get());
+
+            SamplerArray* views = nullptr;
+            if (isFloatingPointFormat(format)||isNormalizedFormat(format)||isScaledFormat(format))
+                views = &m_fsamplers;
+            else if (isSignedFormat(format))
+                views = &m_isamplers;
+            else
+                views = &m_usamplers;
+            auto view_it = std::find_if(views->views->begin(), views->views->end(), [format](const core::smart_refctd_ptr<ICPUImageView>& _view) {return _view->getCreationParameters().format==format;});
+            if (view_it==views->views->end()) //no physical page texture view/sampler for requested format
+                return STextureData::invalid();
+            smplrIndex = std::distance(views->views->begin(), view_it);
+        }
+        auto assignedLayers = storage->getPageTableLayers();
+
+        page_tab_offset_t pgtOffset = page_tab_offset_invalid();
+        for (auto it = assignedLayers.first; it != assignedLayers.second; ++it)
+        {
+            pgtOffset = alloc(_img, _subres, *it);
+            if ((pgtOffset==page_tab_offset_invalid()).all())
+                continue;
+        }
+        if ((pgtOffset==page_tab_offset_invalid()).all())
+        {
+            if (m_freePageTableLayerIDs.empty())
+                return STextureData::invalid();
+            const uint32_t pgtLayer = m_freePageTableLayerIDs.top();
+            m_freePageTableLayerIDs.pop();
+            pgtOffset = alloc(_img, _subres, pgtLayer);
+            if ((pgtOffset==page_tab_offset_invalid()).all())//this would be super weird but let's check
+                return STextureData::invalid();
+            storage->addPageTableLayer(pgtLayer);
+
+            (*m_layerToViewIndexMapping)[pgtLayer] = smplrIndex;
+        }
+
+        const auto extent = _img->getCreationParameters().extent;
+
+        const uint32_t levelsTakingAtLeastOnePageCount = countLevelsTakingAtLeastOnePage(extent, _subres);
+        const uint32_t levelsToPack = std::min(_subres.levelCount, m_pageTable->getCreationParameters().mipLevels+m_pgSzxy_log2);
+
+        uint32_t miptailPgAddr = SPhysPgOffset::invalid_addr;
+
+        using phys_pg_addr_alctr_t = ICPUVTResidentStorage::phys_pg_addr_alctr_t;
+        //TODO up to this line, it's kinda common code for CPU and GPU, refactor later
```

Comment:
the `core::address_allocator_traits<phys_pg_addr_alctr_t>::multi_alloc_addr(storage->tileAlctr` is also common code, you know how large the biggest texture will be so you can preallocate temporary arrays as members large enough to hold the output, size and alignment (for tiles size==alignment==1)

Furthermore you can first try and allocate the physical tiles, then if you fail you only need to free them and return, unlike now when its quite hard and laborious to clean up in event of a failure (you avoid copying over image data if you've failed to allocate).